contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,203 @@
+/**
+
+## The Flow Non-Fungible Token standard
+
+## `NonFungibleToken` contract interface
+
+The interface that all Non-Fungible Token contracts should conform to.
+If a user wants to deploy a new NFT contract, their contract would need
+to implement the NonFungibleToken interface.
+
+Their contract must follow all the rules and naming
+that the interface specifies.
+
+## `NFT` resource
+
+The core resource type that represents an NFT in the smart contract.
+
+## `Collection` Resource
+
+The resource that stores a user's NFT collection.
+It includes a few functions to allow the owner to easily
+move tokens in and out of the collection.
+
+## `Provider` and `Receiver` resource interfaces
+
+These interfaces declare functions with some pre and post conditions
+that require the Collection to follow certain naming and behavior standards.
+
+They are separate because it gives the user the ability to share a reference
+to their Collection that only exposes the fields and functions in one or more
+of the interfaces. It also gives users the ability to make custom resources
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import ViewResolver from "ViewResolver"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+access(all) contract NonFungibleToken {
+
+    // An entitlement for allowing the withdrawal of tokens from a Vault
+    access(all) entitlement Withdrawable
+
+    /// Event that is emitted when a token is withdrawn,
+    /// indicating the owner of the collection that it was withdrawn from.
+    ///
+    /// If the collection is not in an account's storage, `from` will be `nil`.
+    ///
+    access(all) event Withdraw(id: UInt64, uuid: UInt64, from: Address?, type: String)
+
+    access(self) view fun emitNFTWithdraw(id: UInt64, uuid: UInt64, from: Address?, type: String): Bool
+    {
+        emit Withdraw(id: id, uuid: uuid, from: from, type: type)
+        return true
+    }
+
+    /// Event that emitted when a token is deposited to a collection.
+    ///
+    /// It indicates the owner of the collection that it was deposited to.
+    ///
+    access(all) event Deposit(id: UInt64, uuid: UInt64, to: Address?, type: String)
+
+    access(self) view fun emitNFTDeposit(id: UInt64, uuid: UInt64, to: Address?, type: String): Bool
```

Comment:
Should these method take a reference to the resource and just get the fields instead like id/uuid/type? that way it is not possible to lie about those i guess?

Reply:
it isn't possible to lie about the arguments because they get their values from the function arguments of `deposit` and withdraw

Reply:
But somebody else can call them with fake data? Granted they can do that with fake address too?

Reply:
The `emit` functions are `access(self)` so only functions in the standard can call them